packages/client-config/src/generate_client_config.ts

```diff
@@ -18,14 +21,15 @@ import {
  */
 export const generateClientConfig = async (
   credentialProvider: AwsCredentialIdentityProvider,
-  backendIdentifier: DeployedBackendIdentifier
+  backendIdentifier: DeployedBackendIdentifier,
+  version: ClientConfigVersion
 ): Promise<ClientConfig> => {
```

Comment:
Would it be possible to infer narrowed down return type from `version` parameter ?

Reply:
The method return type has to be a union since the value of version is only available at runtime :(. Consumers can however use typeguards with a bunch of if-else to type it to the narrow type.

Reply:
would this work if `version` is enumerated https://stackoverflow.com/questions/54165536/typescript-function-return-type-based-on-input-parameter ?

Reply:
nice! conditional types :)
However I still don't see how it helps narrowing the types as the problem now just shifted to the consumer, i.e. unless they have the constant version passed in, they can't see the narrow type and it will be a union.

Tried it here https://github.com/aws-amplify/amplify-backend/pull/1057/commits/75c04e3a3c1083fa581a7ccdccbe7987d96106da

Reply:
While it doesn't solve the problem 100% it does help in cases you mention - when version is known upfront and is const.

Btw. this makes me think - should we default version to "latest" in this API ? In that case calls without specifying it would default to narrowed type for latest right ?

Reply:
The default is TBD, most likely it will be legacy version. 

Moreover I preferred setting default at the command level (in the cli package) rather than defaulting in the config generation, meaning the default value is still a dynamic variable for this API.

Reply:
I was thinking about programatic usage of this API (outside of CLI context).

Reply:
I.e. default for CLI doesn't have to be the same as programmatic default.